types/time_test.go

```diff
@@ -54,22 +97,26 @@ func (s *testTimeSuite) TestDateTime(c *C) {
 		{"121231113045", "2012-12-31 11:30:45"},
 		{"2012-02-29", "2012-02-29 00:00:00"},
 		{"00-00-00", "0000-00-00 00:00:00"},
-		{"00-00-00 00:00:00.123", "2000-00-00 00:00:00.123"},
+		{"00-00-00 00:00:00.123", "2000-00-00 00:00:00.123000"},
 		{"11111111111", "2011-11-11 11:11:01"},
-		{"1701020301.", "2017-01-02 03:01:00"},
-		{"1701020304.1", "2017-01-02 03:04:01.0"},
-		{"1701020302.11", "2017-01-02 03:02:11.00"},
+		{"1701020304.1", "2017-01-02 03:04:01"},
```

Comment:
Why the test case is changed, what's the result of MySQL for this?

Reply:
`

mysql> select @@sql_mode;
+------------------------+
| @@sql_mode             |
+------------------------+
| NO_ENGINE_SUBSTITUTION |
+------------------------+
1 row in set (0.00 sec)

mysql> Select cast('1701020301.' as datetime);
+---------------------------------+
| cast('1701020301.' as datetime) |
+---------------------------------+
| 2017-01-02 03:01:00             |
+---------------------------------+
1 row in set (0.00 sec)
`

Reply:
mysql> Select cast('1701020302.11' as datetime);
+-----------------------------------+
| cast('1701020302.11' as datetime) |
+-----------------------------------+
| 2017-01-02 03:02:11               |
+-----------------------------------+
1 row in set (0.00 sec)

Reply:
mysql> Select cast('1701020302.11' as datetime);
+-----------------------------------+
| cast('1701020302.11' as datetime) |
+-----------------------------------+
| 2017-01-02 03:02:11               |
+-----------------------------------+
1 row in set (0.00 sec)

Reply:
mysql> Select cast('00-00-00 00:00:00.123' as datetime);
+-------------------------------------------+
| cast('00-00-00 00:00:00.123' as datetime) |
+-------------------------------------------+
| 2000-00-00 00:00:00                       |
+-------------------------------------------+
1 row in set (0.00 sec)

mysql> Select cast('170102037.11' as datetime);
+----------------------------------+
| cast('170102037.11' as datetime) |
+----------------------------------+
| 2017-01-02 03:07:11              |
+----------------------------------+
1 row in set (0.01 sec)

mysql> Select cast('2011-11-11x' as datetime);
+---------------------------------+
| cast('2011-11-11x' as datetime) |
+---------------------------------+
| 2011-11-11 00:00:00             |
+---------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> Select cast('2011-11-11  10:10:10' as datetime);
+------------------------------------------+
| cast('2011-11-11  10:10:10' as datetime) |
+------------------------------------------+
| 2011-11-11 10:10:10                      |
+------------------------------------------+
1 row in set (0.00 sec)

mysql> show variables like '%version%';
+-------------------------+------------------------------+
| Variable_name           | Value                        |
+-------------------------+------------------------------+
| innodb_version          | 5.7.19                       |
| protocol_version        | 10                           |
| slave_type_conversions  |                              |
| tls_version             | TLSv1,TLSv1.1                |
| version                 | 5.7.19-log                   |
| version_comment         | MySQL Community Server (GPL) |
| version_compile_machine | x86_64                       |
| version_compile_os      | Linux                        |
+-------------------------+------------------------------+
8 rows in set (0.00 sec)

Reply:
So, what' happen when sql_mode NO_ENGINE_SUBSTITUTION is not set?
It there a FSP part ?